### PR TITLE
solves sudoku puzzles

### PR DESCRIPTION
**Pull request by TreverArnold**

### CODE REVIEW

File: sudoku.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+def sudoku(puzzle):
+    def is_valid(num, row, col):
```
Comment (W-Ely) [inline sudoku.py:2]: There is usually not a need to nest functions like this, can you pull them out so they are all in the global scope.

File: sudoku.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    for i in range(9):
+        if puzzle[i][col] == num:
+            return False
+
+    rowstart = 3 * (row // 3)
+    colstart = 3 * (col // 3)
+
+    for r in range(rowstart, rowstart + 3):
+        for c in range(colstart, colstart + 3):
+            if puzzle[r][c] == num:
```
Comment (W-Ely) [inline sudoku.py:14]: usually single letter variable names are frowned upon, but in this case it's pretty clear how they are being used.

File: sudoku.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    return True
+
+def solve(puzzle):
+    for row in range(9):
+        for col in range(9):
+            if puzzle[row][col] == 0:
+                for num in range(1, 10):
+                    if is_valid(num, row, col, puzzle):
+                        puzzle[row][col] = num
+                        if solve(puzzle):
```
Comment (W-Ely) [inline sudoku.py:26]: Nice recursive solution. In my opinion this is the easiest way to solve this kind of problem.

In what case can you _not_ use recursion?

**Review by W-Ely — APPROVED** [review]

File: sudoku.py (hunk 3 above)
Comment (TreverArnold) [reply sudoku.py:26]: I know there are cases where recursion may not be the best option, however in terms of completely being unable to use it, I can not think of any.

Comment (W-Ely) [reply sudoku.py:26]:
```
def find_max_depth(n=0):
    print(f"Depth: {n}")
    return find_max_depth(n + 1)
```

**Pull request merged**